api/src/main/java/ca/bc/gov/educ/api/trax/config/RestErrorHandler.java

```diff
@@ -1,63 +1,106 @@
+import static org.springframework.http.HttpStatus.NOT_FOUND;
+
+@Slf4j
 @ControllerAdvice
 public class RestErrorHandler extends ResponseEntityExceptionHandler {
 
-	private static final Logger LOGGER = Logger.getLogger(RestErrorHandler.class);
+	GradValidation validation;
 
 	@Autowired
-	GradValidation validation;
+	public RestErrorHandler(GradValidation validation) {
+		this.validation = validation;
+	}
+
+	/**
+	 * Override method handles exception thrown by @Valid validation
+	 * on controller methods (for example)
+	 */
+	@Override
+	protected ResponseEntity<Object> handleMethodArgumentNotValid(MethodArgumentNotValidException ex, HttpHeaders headers, HttpStatusCode status, WebRequest request) {
+		Map<String, String> errors = new HashMap<>();
+		ex.getBindingResult().getAllErrors().forEach(error -> {
+			String fieldName = ((FieldError) error).getField();
+			String errorMessage = error.getDefaultMessage();
+			errors.put(fieldName, errorMessage);
+		});
+		ApiError apiError = new ApiError(BAD_REQUEST);
+		apiError.setMessage(errors.toString());
+		return buildResponseEntity(apiError);
+	}
+
+	/**
+	 * Handles the exception thrown by not found and translates to 404 response
+	 * @param ex the exception
+	 * @return a 404 with message
+	 */
+	@ExceptionHandler(value = { EntityNotFoundException.class })
+	protected ResponseEntity<Object> handleEntityNotFoundException(EntityNotFoundException ex){
+		ApiError apiError = new ApiError(NOT_FOUND);
+		apiError.setMessage(ex.getMessage());
+		return buildResponseEntity(apiError);
+	}
 
 	@ExceptionHandler(value = { IllegalArgumentException.class, IllegalStateException.class })
-	protected ResponseEntity<Object> handleConflict(RuntimeException ex, WebRequest request) {
-		LOGGER.error("Illegal argument ERROR IS: " + ex.getClass().getName(), ex);
+	protected ResponseEntity<Object> handleConflict(RuntimeException ex) {
+        log.error("Illegal argument ERROR IS: {}", ex.getClass().getName(), ex);
 		ApiResponseModel<?> response = ApiResponseModel.ERROR(null, ex.getLocalizedMessage());
-		validation.ifErrors(errorList -> response.addErrorMessages(errorList));
-		validation.ifWarnings(warningList -> response.addWarningMessages(warningList));
+		validation.ifErrors(response::addErrorMessages);
+		validation.ifWarnings(response::addWarningMessages);
 		validation.clear();
 		return new ResponseEntity<>(response, HttpStatus.UNPROCESSABLE_ENTITY);
 	}
 
 	@ExceptionHandler(value = { JpaObjectRetrievalFailureException.class, DataRetrievalFailureException.class })
-	protected ResponseEntity<Object> handleEntityNotFound(RuntimeException ex, WebRequest request) {
-		LOGGER.error("JPA ERROR IS: " + ex.getClass().getName(), ex);
+	protected ResponseEntity<Object> handleEntityNotFound(RuntimeException ex) {
+        log.error("JPA ERROR IS: {}", ex.getClass().getName(), ex);
 		validation.clear();
 		return new ResponseEntity<>(ApiResponseModel.ERROR(null, ex.getLocalizedMessage()), HttpStatus.BAD_REQUEST);
 	}
```

Comment:
This is the second EntityNotFound handler, but covers DataRetrievalFailure. 
Also DataRetrievalFailure is more related to the server side 500 group of errors and returning of BAD_REQUEST is confusing.

Reply:
Agree. I merged the two and return a 404. The EntityNotFoundException is the only one explicitly thrown to state that an object is not found in the DB, while the other exceptions are from Spring JPA but also denote not found.